vision_doc.md

```diff
@@ -0,0 +1,279 @@
+# Vision
+`derive_builder` provides an ergonomic way to deal with construction of guaranteed-complete objects from a range of inputs:
+
+* Statically-known values
+* Function arguments
+* Wire/file/database inputs
+
+The builder should be focused specifically on the assembly of inputs and validating the completeness of the parameter set; additional validation should be handled by a constructor function provided by the target type. That constructor function may be public or private, depending on whether or not the crate author wants to _require_ the use of the builder.
```

Comment:
I've moved it to a sub-section, but I do think that having some level of opinion on where validation fits in is significant. (Also, not sure where that statement would go otherwise as the doc stands now).